src/app/utilities/constraint-utilities/constraint-copier.ts

```diff
@@ -0,0 +1,111 @@
+import {AbstractConstraintVisitor} from './abstract-constraint-visitor';
+import {ValueConstraint} from '../../models/constraint-models/value-constraint';
+import {TrueConstraint} from '../../models/constraint-models/true-constraint';
+import {TrialVisitConstraint} from '../../models/constraint-models/trial-visit-constraint';
+import {TimeConstraint} from '../../models/constraint-models/time-constraint';
+import {SubjectSetConstraint} from '../../models/constraint-models/subject-set-constraint';
+import {StudyConstraint} from '../../models/constraint-models/study-constraint';
+import {PedigreeConstraint} from '../../models/constraint-models/pedigree-constraint';
+import {NegationConstraint} from '../../models/constraint-models/negation-constraint';
+import {ConceptConstraint} from '../../models/constraint-models/concept-constraint';
+import {CombinationConstraint} from '../../models/constraint-models/combination-constraint';
+import {Constraint} from '../../models/constraint-models/constraint';
+
+/**
+ * Constraint copy class, for creating a copy of a constraint that can be
+ * edited independently.
+ */
+export class ConstraintCopier extends AbstractConstraintVisitor<Constraint> {
```

Comment:
This could be done by polymorphism for the Constraint class and its descendants, e.g. a method named `copy()`, so that the copying is done generically and one does not need to call these different visitor functions for different types of constraints

Reply:
Yes, indeed, that is another approach that would work. It would make the constraint classes a bit heavy in my opinion. E.g., also serialisation could be approached that way, but then you have:
- the serialisation/copying code scattered over several classes
- the constraint classes carrying many responsibilities (capture properties, cloning, serlalisation, ...)

Reply:
serialisation indeed needs to be separated as it is about converting glowing bear constraints to a backend-specific format, whereas the copy() method is another story.

Reply:
Moved to the Constraint classes.